targets/TARGET_Analog_Devices/TARGET_ADUCM4X50/TARGET_ADUCM4050/api/trng_api.c

```diff
@@ -0,0 +1,96 @@
+/*******************************************************************************
+ * Copyright (c) 2010-2017 Analog Devices, Inc.
+ *
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without modification,
+ * are permitted provided that the following conditions are met:
+ *   - Redistributions of source code must retain the above copyright notice,
+ *     this list of conditions and the following disclaimer.
+ *   - Redistributions in binary form must reproduce the above copyright notice,
+ *     this list of conditions and the following disclaimer in the documentation
+ *     and/or other materials provided with the distribution.
+ *   - Modified versions of the software must be conspicuously marked as such.
+ *   - This software is licensed solely and exclusively for use with processors
+ *     manufactured by or for Analog Devices, Inc.
+ *   - This software may not be combined or merged with other code in any manner
+ *     that would cause the software to become subject to terms and conditions
+ *     which differ from those listed here.
+ *   - Neither the name of Analog Devices, Inc. nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *   - The use of this software may or may not infringe the patent rights of one
+ *     or more patent holders.  This license does not release you from the
+ *     requirement that you obtain separate licenses from these patent holders
+ *     to use this software.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES, INC. AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-
+ * INFRINGEMENT, TITLE, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL ANALOG DEVICES, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, PUNITIVE OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, DAMAGES ARISING OUT OF
+ * CLAIMS OF INTELLECTUAL PROPERTY RIGHTS INFRINGEMENT; PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ ******************************************************************************/
+
+#if defined(DEVICE_TRNG)
+
+#include <stdlib.h>
+#include <adi_rng.h>
+#include <adi_pwr.h>
+#include "cmsis.h"
+#include "trng_api.h"
+
+// Sampling counter values
+// Prescaler: 0 - 10
+// LenReload: 0 - 4095
+#define TRNG_CNT_VAL	4095
+#define TRNG_PRESCALER 	2
```

Comment:
what does the values of `TRNG_CNT_VAL` and `TRNG_PRESCALER` defined? What do they mean? Do these values gurantee there will be full entropy from the source?

Reply:
These set the sample length register in the RNG hardware. The sample length register determines the number of samples to accumulate in a CRC register when generating a random number.

Reply:
I see, thanks, so isn't this dependent on the value of  `length` in `trng_get_bytes` ?
What if the required `length` will be larger than the total sample length defined in compilation?

Reply:
The sample length register value in the HW is not the same as the length you pass into trng_get_bytes. For each random data byte trng_get_bytes requests from the HW, the HW would need to accumulate X number of samples (in this case it uses values derived from clock jitter) before outputting a value, and the sample length register sets this X value. That's why the driver has that bReady flag to indicate whether the accumulation is complete.  TRNG_CNT_VAL and TRNG_PRESCALER together define this X value.

Reply:
I'm sorry, but it's still not clear to me. X number of samples is a constant value, right? I mean, let's say every sample is 32 bit, and X would be 5, this means that every time you ask for random value, you will generate only 160 bits of entropy. Wht if the required random length is 180 bits? If I understand correct, the the `bready` is set once 160 bits were accumulated, right?

Reply:
The value of X is not the number of entropy bits you want to generate. X is the number of sample accumulations the RNG hardware runs internally before outputting any entropy bits (the samples come from the clock jitter measurements). The bReady flag indicates the completion of this accumulation process, so when the flag is set, in the current configuration the RNG hardware returns 8-bits of entropy. In order to get 160 bits, the for loop in the trng_get_bytes function will need to loop through 20 times (the bReady flag will need to be set 20 times during this process), and for 180 bits, the loop will need to go through 23 iterations. 

I am not sure whether I have answered your question. If you want to know how the hardware generates the entropy, you can have a look at chapter 13 of the processor reference manual:
http://www.analog.com/media/en/dsp-documentation/processor-manuals/ADuCM4050_HRM_0.2.pdf

Reply:
Thanks, that answers my question